### PR TITLE
Df new dg assembly

### PR DESCRIPTION
**Pull request by dflanderka**
- improve class AssemblyDG

### CODE REVIEW

File: src/transport/transport_dg.cc [hunk 1]
```diff
@@ -205,16 +206,18 @@ void TransportDG<Model>::EqData::set_DG_parameters_boundary(Side side,
 template<class Model>
 TransportDG<Model>::TransportDG(Mesh & init_mesh, const Input::Record in_rec)
         : Model(init_mesh, in_rec),
+		  data_( make_shared<EqData>() ),
           input_rec(in_rec),
-          allocation_done(false)
+          allocation_done(false),
+		  multidim_assembly_new_{AssemblyDGNew<1, Model>(data_, *this), AssemblyDGNew<2, Model>(data_, *this), AssemblyDGNew<3, Model>(data_, *this)}
```
Comment (dflanderka) [inline src/transport/transport_dg.cc]: Initialization provides correct inner AssemblyDGNew objects only in this constructor. In other methods these objects return unexpected results.

File: src/transport/assembly_dg.hh [hunk 2]
```diff
@@ -795,6 +1177,10 @@ private:
         model_.velocity_field_ptr()->value_list(point_list, cell, velocity);
     }
 
+    std::shared_ptr<BulkIntegral> bulk_integral_;          ///< Bulk integrals of elements of given dimension
```
[the remark this answers is not in the document]
Comment (jbrezmorf) [reply src/transport/assembly_dg.hh]: My fault. *Integral objects are necessary. Keep them.
However, the contain just small data so no shared pointers should be necessary. We can consider:
- Making them just structs with some helpful methods. Without any explicit constructor, the default copy and assignment should be available.
- Possibly merge them in a single object and also have a single creation method in eval_points.

... can be done later.

File: src/transport/assembly_dg.hh [hunk 3]
```diff
@@ -25,137 +25,434 @@
...
+
+        if (fe_values_!=nullptr) delete fe_values_;
+
+    }
+
+    void create_integrals(std::shared_ptr<EvalPoints> eval_points) {
+        bulk_integral_ = eval_points->add_bulk<dim>(*quad_);
+        //edge_integral_ = eval_points->add_edge<dim>(*quad_low_);
+        //if (dim>1) coupling_integral_ = eval_points->add_coupling<dim>(*quad_low_);
+        //boundary_integral_ = eval_points->add_boundary<dim>(*quad_low_);
```
Comment (dflanderka) [inline src/transport/assembly_dg.hh]: Remove dependency in GenericAssembly::add_xyz_integral(), delete unused integral objects. Same in other descendants of AssemblyBase.

File: src/transport/assembly_dg.hh [hunk 4]
```diff
@@ -761,22 +1544,182 @@ public:
...
+
+        friend class TransportDG<Model>;
+        template < template<Dim...> class DimAssembly>
+        friend class GenericAssembly;
+
+};
+
+
+template <unsigned int dim, class Model>
+class AssemblyDG : public AssemblyBase<dim>
```
Comment (dflanderka) [inline src/transport/assembly_dg.hh]: Remove class.

File: src/transport/transport_dg.hh [hunk 5]
```diff
@@ -210,6 +222,13 @@ public:
         /// Object for distribution of dofs.
         std::shared_ptr<DOFHandlerMultiDim> dh_;
 
+        /// general assembly objects, hold assembly objects of appropriate dimension
+        GenericAssembly< AssemblyDGDim > * generic_assembly_;
```
Comment (dflanderka) [inline src/transport/transport_dg.hh]: Replace and remove.

File: src/transport/transport_dg.cc [hunk 6]
```diff
@@ -330,10 +357,27 @@ void TransportDG<Model>::initialize()
...
 
     // initialization of balance object
-    Model::balance_->allocate(data_->dh_->distr()->lsize(),
-            max(assembly1_->fe_->n_dofs(), max(assembly2_->fe_->n_dofs(), assembly3_->fe_->n_dofs())));
-
-    initialize_assembly_objects();
+    Model::balance_->allocate(data_->dh_->distr()->lsize(), data_->generic_assembly_->eval_points()->max_size());
+
+    // initialization of assembly object
+    data_->generic_assembly_->multidim_assembly().get<1>()->initialize(*this);
```
Comment (dflanderka) [inline src/transport/transport_dg.cc]: Move initialization to specialized method of GenericAssembly.

File: src/transport/transport_dg.cc [hunk 7]
```diff
@@ -735,17 +734,11 @@ void TransportDG<Model>::set_initial_condition()
     START_TIMER("set_init_cond");
```
Comment (dflanderka) [inline src/transport/transport_dg.cc]: Solve two calling of assemble method, remove this method.

File: src/transport/transport_dg.cc [hunk 8]
```diff
@@ -237,6 +238,11 @@ TransportDG<Model>::TransportDG(Mesh & init_mesh, const Input::Record in_rec)
 	multidim_assembly_.push_back( std::dynamic_pointer_cast<AssemblyDGBase>(assembly1_) );
 	multidim_assembly_.push_back( std::dynamic_pointer_cast<AssemblyDGBase>(assembly2_) );
 	multidim_assembly_.push_back( std::dynamic_pointer_cast<AssemblyDGBase>(assembly3_) );
+	std::shared_ptr<AssemblyDGNew<1, Model>> assembly_new1 = std::make_shared<AssemblyDGNew<1, Model>>(data_, *this);
+	std::shared_ptr<AssemblyDGNew<2, Model>> assembly_new2 = std::make_shared<AssemblyDGNew<2, Model>>(data_, *this);
+	std::shared_ptr<AssemblyDGNew<3, Model>> assembly_new3 = std::make_shared<AssemblyDGNew<3, Model>>(data_, *this);
+	multidim_assembly_new_ = MultidimAssemblyDGNew<Model>(assembly_new1, assembly_new2, assembly_new3);
```
[the remark this answers is not in the document]
Comment (jbrezmorf) [reply src/transport/transport_dg.cc]: Proposed changes in mixed.hh to have the support of mixed templates starting from dim 1.
file:///home/jb/workspace/flow123d/src/tools/mixed.hh

File: src/transport/transport_dg.cc (hunk 6 above)
Comment (dflanderka) [reply src/transport/transport_dg.cc]: It is not possible, 'this' object depends on Model.

**Pull request merged**